### PR TITLE
Return an error in strict mode if yaml contains duplicate values for a map or struct

### PR DESCRIPTION
**Pull request by aeijdenberg**
Return an error in strict mode if yaml contains duplicate values for a map or struct.

Fixes <https://github.com/go-yaml/yaml/issues/284>.

### CODE REVIEW

File: decode.go [hunk 1]
```diff
@@ -625,7 +635,17 @@ func (d *decoder) mappingStruct(n *node, out reflect.Value) (good bool) {
 		if !d.unmarshal(ni, name) {
 			continue
 		}
-		if info, ok := sinfo.FieldsMap[name.String()]; ok {
+		ns := name.String()
+		if d.strict {
+			for _, df := range doneFields {
```
Comment (rogpeppe) [inline decode.go]: This linear search seems a little more heavyweight than it needs to be. If we keep a unique field index in each fieldInfo (the index into fieldsList could be used when building up the struct info in getStructInfo, perhaps), then doneFields could be a []bool, and checking whether a field has already been done would just require indexing into the slice.

File: decode.go [hunk 2]
```diff
@@ -625,7 +635,17 @@ func (d *decoder) mappingStruct(n *node, out reflect.Value) (good bool) {
...
+				if ns == df {
+					d.terrors = append(d.terrors, fmt.Sprintf("line %d: field %s already set in struct %s", n.line+1, ns, out.Type()))
```
Comment (rogpeppe) [inline decode.go]: perhaps `set in type %s` so that we don't stutter when the type is an unnamed struct?

File: decode_test.go [hunk 3]
```diff
@@ -977,14 +978,21 @@ func (s *S) TestUnmarshalSliceOnPreset(c *C) {
...
 	c.Check(err, IsNil)
 	err = yaml.Unmarshal([]byte("a: 1\nb: 2\nc: 3"), &v)
 	c.Check(err, IsNil)
 	err = yaml.UnmarshalStrict([]byte("a: 1\nb: 2\nc: 3"), &v)
-	c.Check(err, ErrorMatches, "yaml: unmarshal errors:\n  line 1: field c not found in struct struct { A int; B int }")
+	c.Check(err, ErrorMatches, "yaml: unmarshal errors:\n  line 1: field c not found in struct struct { A int; B int; Z map\\[string\\]string }")
+	err = yaml.UnmarshalStrict([]byte("a: 1\na: 2"), &v)
+	c.Check(err, ErrorMatches, "yaml: unmarshal errors:\n  line 1: field a already set in struct struct { A int; B int; Z map\\[string\\]string }")
+	err = yaml.UnmarshalStrict([]byte("z:\n  a: 1\n  a: 1"), &v)
+	c.Check(err, ErrorMatches, "yaml: unmarshal errors:\n  line 2: field a already set in map")
```
Comment (rogpeppe) [inline decode_test.go]: Let's have a test for unmarshaling duplicate values into a struct with inline members too, please (including an inline struct and a map)

File: decode.go [hunk 4]
```diff
@@ -507,6 +507,15 @@ func (d *decoder) sequence(n *node, out reflect.Value) (good bool) {
 	return true
 }
 
+func (d *decoder) setMapIndex(n *node, out, k, v reflect.Value) {
+	if d.strict {
+		if out.MapIndex(k) != zeroValue {
+			d.terrors = append(d.terrors, fmt.Sprintf("line %d: field %s already set in map", n.line+1, k.String()))
```
Comment (rogpeppe) [inline decode.go:544]: Shouldn't this return without setting the value? (not that it matters that much I guess)

**Review by rogpeppe — CHANGES_REQUESTED** [review]
Thanks very much for this. This looks reasonable, but I'd prefer to use a lighter weight solution for keeping track of already used struct fields. Also, this needs an update to the UnmarshalStrict docs to document the new behaviour.

File: decode.go (hunk 4 above)
Comment (aeijdenberg) [reply decode.go:544]: While it probably doesn't matter much, I agree it shouldn't set a value after it's decided that's an error. Fixing as suggested.

File: decode.go (hunk 1 above)
Comment (aeijdenberg) [reply decode.go]: I nearly followed this, but got a bit lost trying to understand the `InlineMap` stuff...

Instead I changed `doneFields` to a `map[string]bool` - which I think at minimum makes the code easier to read.

File: decode.go (hunk 2 above)
Comment (aeijdenberg) [reply decode.go]: Fixed.

File: decode_test.go (hunk 3 above)
Comment (aeijdenberg) [reply decode_test.go]: Added, I think - I hadn't seen this form before so not 100% confident.

**Comment by jsha** [comment]
@rogpeppe do you have time for a re-review? On the Let's Encrypt Boulder project we'd very much like to enable detection of duplicate keys in YAML maps. Thanks!

**Comment by jsha** [comment]
@aeijdenberg it looks like this has conflicts now; can you merge master and resolve the conflicts?

**Comment by aeijdenberg** [comment]
@jsha - I've just rebased and squashed on the latest `devel` (I think that's the active development branch).

**Comment by rogpeppe** [comment]
Superceded by PR #307. Thanks very much for this anyway - it pointed me in exactly the right direction.

**Pull request closed without merging**